modules/ggl-sdk/src/object.c

```diff
@@ -3,105 +3,171 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include <assert.h>
+#include <ggl/attr.h>
 #include <ggl/buffer.h>
+#include <ggl/log.h>
 #include <ggl/object.h>
 #include <string.h>
-#include <stdalign.h>
 #include <stdbool.h>
 #include <stdint.h>
 
-typedef struct {
-    union {
-        bool boolean;
-        int64_t i64;
-        double f64;
-        GglBuffer buf;
-        GglList list;
-        GglMap map;
-    };
-
-    uint8_t type;
-} GglObjectPriv;
-
 static_assert(
-    sizeof(GglObject) == sizeof(GglObjectPriv), "GglObject impl invalid."
+    sizeof(void *) == sizeof(size_t),
+    "Platforms where pointers and size_t are not same width are not currently "
+    "supported."
 );
+
 static_assert(
-    alignof(GglObject) == alignof(GglObjectPriv), "GglObject impl invalid."
+    (sizeof(size_t) == 4) || (sizeof(size_t) == 8),
+    "Platform not 32 or 64 bits?"
 );
 
-static GglObject obj_from_priv(GglObjectPriv obj) {
-    GglObject result;
-    memcpy(&result, &obj, sizeof(GglObject));
-    return result;
-}
-
-static GglObjectPriv priv_from_obj(GglObject obj) {
-    GglObjectPriv result;
-    memcpy(&result, &obj, sizeof(GglObject));
-    return result;
-}
-
 GglObjectType ggl_obj_type(GglObject obj) {
-    return (GglObjectType) priv_from_obj(obj).type;
+    // Last byte is tag
+    uint8_t result = obj._private[sizeof(obj._private) - 1];
+    assert(result <= 7);
```

Comment:
The `GglObjectType` enum is 0-indexed, so `GGL_TYPE_MAP` is 6.
```suggestion
    assert(result < 7);
```

Reply:
ah